RHDEVDOCS-3360 - How to run OpenShift GitOps control plane on infra nodes (#40180)
openshift/openshift-docs: modules/go-add-infra-nodes.adoc

// Module included in the following assembly:
//
// * gitops/gitops-release-notes.adoc

[id="add-infra-nodes_{context}"]

> = Adding Infra Nodes

Reviewer: ```suggestion
= Adding infrastructure nodes
```
- Headings should use sentence case
- Spell out `infra` as `infrastructure` throughout this content

.Procedure

> . Label the existing nodes as Infrastructure usinf the following command:app-name:

Reviewer: ```suggestion
. Label existing nodes as `Infrastructure` using the following command:app-name:
```

+
[source,terminal]
----
$ oc label node <node-name> node-role.kubernetes.io/infra
----

> . Add `runOnInfra` toggle in the `GitOpsService` Custom Resource (CR) to add infra node selector:

Reviewer: ```suggestion
. Add the `runOnInfra` toggle to the `GitOpsService` Custom Resource (CR) to add the infrastructure node selector:
```

+
[source,terminal]
----
$ oc edit gitopsservice -n openshift-gitops
----

> . Add `runOnInfra` to spec and set to `true` in the YAML:

Reviewer: ```suggestion
. In the ????.yml file, add `runOnInfra` to the `spec` section and set to it to `true`:
```
- It's not clear to me that the first part of this step is telling the user to add a section to the YAML file. Maybe phrase it per my suggestion to make this step more clear.
- Also, indicate what YML file the user should be altering as part of this process.

+
[source,yaml]
----
apiVersion: pipelines.openshift.io/v1alpha1
kind: GitopsService
metadata:
  name: cluster
spec:
  runOnInfra: true
----
. Optional: You can also apply taints and isolate the workloads on infrastructure nodes and prevent other workloads from scheduling on these nodes. 
+
[source,terminal]
----
$ oc adm taint nodes -l node-role.kubernetes.io/infra 
infra=reserved:NoSchedule infra=reserved:NoExecute
----

> If you have applied taints to the nodes, you can add tolerations in the `GitOpsService` CR:app-name:

Reviewer: ```suggestion
If you apply taints to the nodes, you can add tolerations in the `GitOpsService` CR:app-name:
```

+
[source,yaml]
----
spec:
  runOnInfra: true
  tolerations:
  - effect: NoSchedule
    key: infra
    value: reserved
  - effect: NoExecute
    key: infra
    value: reserved 
----

> To verify that the workloads are scheduled on infrastructure nodes in the {gitops-title} namespace, click on any of the pods and see the details. The *nodeSelector* and the *tolerations* will be added.

Reviewer: ```suggestion
To verify that the workloads are scheduled on infrastructure nodes in the {gitops-title} namespace, click any of the pod names and ensure that *nodeSelector* and  *tolerations* have been added.
```
- Per IBM SG: do not use `click on`; use "click" instead
- Change `pod names` instead of `pods`, depending on what the user actually has to click in the UI.

// Module included in the following assembly:
//

> // * gitops/gitops-release-notes.adoc

Reviewer 2: Change the assembly name to the actual assembly.

= Adding infrastructure nodes

.Procedure

. Label existing nodes as infrastructure by running the following command:
+
[source,terminal]
----
$ oc label node <node-name> node-role.kubernetes.io/infra
----
. Edit the `GitOpsService` Custom Resource (CR) to add the infrastructure node selector:
+
[source,terminal]
----
$ oc edit gitopsservice -n openshift-gitops
----
. In the `GitOpsService` CR file, add `runOnInfra` toggle to the `spec` section and set it to `true`:
+
[source,yaml]
----
apiVersion: pipelines.openshift.io/v1alpha1
kind: GitopsService
metadata:
  name: cluster
spec:
  runOnInfra: true
----
. Optional: Apply taints and isolate the workloads on infrastructure nodes and prevent other workloads from scheduling on these nodes. 
+
[source,terminal]
----
$ oc adm taint nodes -l node-role.kubernetes.io/infra 
infra=reserved:NoSchedule infra=reserved:NoExecute
----
+
. Optional: If you apply taints to the nodes, you can add tolerations in the `GitOpsService` CR:app-name:
+
[source,yaml]
----
spec:
  runOnInfra: true
  tolerations:
  - effect: NoSchedule
    key: infra
    value: reserved
  - effect: NoExecute
    key: infra
    value: reserved 
----

> To verify that the workloads are scheduled on infrastructure nodes in the {gitops-title} namespace, click any of the pod names and ensure that the *Node selector* and *Tolerations* have been added.

Reviewer 2: The {gitops-title} attribute does not render in the preview, pls verify that the include for the gitops attribute file has been added to the assembly, looks like it is missing.
![Screenshot from 2022-01-07 19-08-18](https://user-images.githubusercontent.com/16647625/148551951-188831e6-f4d4-4f9e-b86c-5ea888f3d452.png)

[NOTE]
====

> Any manually added *nodeSelectors* and *tolerations* added to the default Argo CD CR will be overwritten by the toggle and the tolerations in the `GitOpsService` CR.

Reviewer 2: This comment has not been fixed @DebarghoGhosh
